### PR TITLE
🦾  525 adding versioning and edit mode for test definitions

### PR DESCRIPTION
**Pull request by xoscar**
This PR adds versioning to the FE and enables users to draft changes until they are ready to publish or if they want to revert all to the previous state.

## Changes

- New edit mode.
- Test header now displays the version.
- Displays a warning if there is a version mismatch.
- New UX for assertion cards.

## Fixes

- https://github.com/kubeshop/tracetest/issues/536
- https://github.com/kubeshop/tracetest/issues/525

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [x] added a test

Screen Shot 2022-05-26 at 11 19 19 a m

### CODE REVIEW

File: web/cypress/integration/Trace/CreateAssertion.spec.ts [hunk 1]
```diff
@@ -123,4 +125,41 @@ describe('Create Assertion', () => {
     cy.get('[data-cy=assertion-card-list]').should('be.visible');
     cy.get('[data-cy=assertion-card]').should('have.lengthOf', 2);
   });
+
+  it('should publish the changes', () => {
```
Comment (xoscar) [inline web/cypress/integration/Trace/CreateAssertion.spec.ts:129]: More tests!

File: web/src/components/AssertionCard/AssertionCard.tsx [hunk 2]
```diff
@@ -42,10 +46,12 @@ const AssertionCard: React.FC<TAssertionCardProps> = ({
           </S.SelectorListText>
           <S.SpanCountText>{spanCountText}</S.SpanCountText>
         </div>
-        <div>
+        <S.ActionsContainer>
```
Comment (xoscar) [inline web/src/components/AssertionCard/AssertionCard.tsx:49]: We are now showing the draft and deleted tags

File: web/src/components/AssertionForm/AssertionFormConfirmModal.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import {Modal} from 'antd';
```
Comment (xoscar) [inline web/src/components/AssertionForm/AssertionFormConfirmModal.tsx:1]: We are now alerting the user that there is a version mismatch before opening the assertion form

File: web/src/components/AssertionForm/AssertionFormProvider.tsx [hunk 4]
```diff
@@ -81,7 +95,18 @@ const AssertionFormProvider: React.FC<{testId: string}> = ({children}) => {
     [isOpen, open, close, formProps, onSubmit, isCollapsed, setIsCollapsed]
   );
 
-  return <Context.Provider value={contextValue}>{children}</Context.Provider>;
+  return (
+    <Context.Provider value={contextValue}>
+      {children}
+      <AssertionFormConfirmModal
```
Comment (xoscar) [inline web/src/components/AssertionForm/AssertionFormProvider.tsx:101]: Adding the confirm modal

File: web/src/components/TestHeader/TestHeader.tsx [hunk 5]
```diff
@@ -8,20 +8,32 @@ interface TTestHeaderProps {
...
+  testVersion,
+}) => {
   const {setIsCollapsed} = useAssertionForm();
+
   return (
     <S.TestHeader>
       <S.Content>
         <S.BackIcon data-cy="test-header-back-button" onClick={onBack} />
-        <S.TestName data-cy="test-details-name">{name}</S.TestName>
+        <S.TestName data-cy="test-details-name">
```
Comment (xoscar) [inline web/src/components/TestHeader/TestHeader.tsx:28]: Versions!

File: web/src/components/TraceActions/TraceActions.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import {Button} from 'antd';
+import {useTestDefinition} from '../../providers/TestDefinition/TestDefinition.provider';
+import * as S from './TraceActions.styled';
+
+const TraceActions: React.FC = () => {
+  const {definitionList, publish, cancel} = useTestDefinition();
+  const pendingCount = definitionList.filter(({isDraft}) => isDraft).length;
```
Comment (xoscar) [inline web/src/components/TraceActions/TraceActions.tsx:7]: New edit actions content.

File: web/src/components/TraceDrawer/TraceDrawerHeader.tsx [hunk 7]
```diff
@@ -54,7 +65,17 @@ const TraceDrawerHeader: React.FC<IProps> = ({
           disabled={isDisabled}
           onClick={event => {
             event.stopPropagation();
-            open();
+            open({
+              isEditing: false,
+              defaultValues: {
+                selectorList:
```
Comment (xoscar) [inline web/src/components/TraceDrawer/TraceDrawerHeader.tsx:71]: opening the assertion form with the selected span details

File: web/src/pages/Trace/Trace.tsx [hunk 8]
```diff
@@ -12,15 +12,15 @@ const TracePage = () => {
 
   return (
     <ReactFlowProvider>
-      <TestDefinitionProvider testId={testId} runId={runId}>
-        <AssertionFormProvider testId={testId}>
-          <TestRunProvider testId={testId} runId={runId}>
+      <TestRunProvider testId={testId} runId={runId}>
```
Comment (xoscar) [inline web/src/pages/Trace/Trace.tsx:15]: Moving the providers as some of them have dependencies

File: web/src/providers/TestDefinition/TestDefinition.provider.tsx [hunk 9]
```diff
@@ -29,36 +45,48 @@ interface IProps {
...
-      await rerun({testId, runId});
-    },
-    [dispatch, rerun, runId, testId]
-  );
+  const {add, cancel, publish, remove, dryRun, update, isDraftMode, init, reset} = useTestDefinitionCrud({
+    testId,
+    runId,
+  });
 
-  const update = useCallback(
```
Comment (xoscar) [inline web/src/providers/TestDefinition/TestDefinition.provider.tsx:43]: Clean up, moving things to hooks and adding side effects

File: web/src/providers/TestDefinition/hooks/useTestDefinitionCrud.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    (assertionResults: TAssertionResults) => {
+      dispatch(initDefinitionList({assertionResults}));
+    },
+    [dispatch]
+  );
+
+  const reset = useCallback(() => {
+    dispatch(resetAction());
+  }, [dispatch]);
+
```
Comment (xoscar) [inline web/src/providers/TestDefinition/hooks/useTestDefinitionCrud.ts:80]: Test Definition Slice proxy actions

File: web/src/redux/actions/TestDefinition.actions.ts [hunk 11]
```diff
@@ -1,53 +1,48 @@
...
+  selector: string;
+  action: 'add' | 'remove' | 'update';
+  patch: PatchCollection;
+};
 
-      const rawDefinitionList = test.definition.definitionList.map(def => TestDefinitionService.toRaw(def));
+export type TCrudResponse = {
+  definitionList: TTestDefinitionEntry[];
+  change: TChange;
+};
```
Comment (xoscar) [inline web/src/redux/actions/TestDefinition.actions.ts:22]: Leaving the async ones here, the rest are moved to the slice

File: web/src/selectors/TestDefinition.selectors.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+import {RootState} from '../redux/store';
+
+const stateSelector = (state: RootState) => state.testDefinition;
+const selectorSelector = (state: RootState, selector: string) => selector;
+
+const selectDefinitionList = createSelector(stateSelector, ({definitionList}) => definitionList);
+
+const TestSelectors = () => ({
+  selectDefinitionList,
+  selectIsLoading: createSelector(stateSelector, ({isLoading}) => isLoading),
```
Comment (xoscar) [inline web/src/selectors/TestDefinition.selectors.ts:11]: Moar selectors

File: web/src/types/TestDefinition.types.ts [hunk 13]
```diff
@@ -6,6 +6,13 @@ export type TRawTestDefinition = TTestSchemas['TestDefinition'];
 export type TTestDefinitionEntry = {
   selector: string;
   assertionList: TAssertion[];
+  isDraft: boolean;
```
Comment (xoscar) [inline web/src/types/TestDefinition.types.ts:9]: Added draft and deleted booleans so we can display them on the cards

**Review by kdhamric — APPROVED** [review]
That is a lot of code!  Skimmed through it - will be great functionality!!

File: web/src/redux/slices/TestDefinition.slice.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+export const assertionResultsToDefinitionList = (assertionResults: TAssertionResults): TTestDefinitionEntry[] => {
+  return assertionResults.resultList.map(({selector, resultList}) => ({
+    isDraft: false,
+    isDeleted: false,
+    selector,
+    assertionList: resultList.flatMap(({assertion}) => [assertion]),
+  }));
+};
+
+const counterSlice = createSlice({
```
Comment (jorgeepc) [inline web/src/redux/slices/TestDefinition.slice.ts:32]: `testDefinitionSlice` ?

**Review by jorgeepc — APPROVED** [review]
🔥

**Pull request merged**